internal/linker/linker.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		panic(fmt.Errorf("cannot retrieve patches info: %v", err))
+	}
+	linkerPatchesVer = tmpVer
+	linkerPatches = tmpPatch
+}
+
+func loadPatches(patchesFs fs.FS) (string, map[string]string, error) {
+	versionHash := sha256.New()
+	patches := make(map[string]string)
+	err := fs.WalkDir(patchesFs, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		f, err := patchesFs.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		var patchBuf bytes.Buffer
+		if _, err := io.Copy(&patchBuf, f); err != nil {
+			return err
+		}
+
+		patchBytes := patchBuf.Bytes()
+
+		if _, err := versionHash.Write(patchBytes); err != nil {
+			return err
+		}
+
+		files, _, err := gitdiff.Parse(bytes.NewReader(patchBytes))
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if file.IsDelete || file.IsRename {
+				panic("delete and rename patch not supported")
+			}
+
+			patches[file.OldName] = string(patchBytes)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return "", nil, err
+	}
+	return base64.RawStdEncoding.EncodeToString(versionHash.Sum(nil)), patches, nil
+}
+
+func applyPatch(workingDirectory, patch string) error {
+	cmd := exec.Command("git", "-C", workingDirectory, "apply")
+	cmd.Stdin = strings.NewReader(patch)
+	return cmd.Run()
+}
+
+func copyFile(src, target string) error {
+	targetDir := filepath.Dir(target)
+	if err := os.MkdirAll(targetDir, os.ModePerm); err != nil {
+		return err
+	}
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+
+	targetFile, err := os.Create(target)
+	if err != nil {
+		return err
+	}
+	defer targetFile.Close()
+	_, err = io.Copy(targetFile, srcFile)
+	return err
+}
+
+func existsFile(path string) bool {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return false
+	}
+	return !stat.IsDir()
+}
+
+func applyPatches(srcDirectory, workingDirectory string) (map[string]string, error) {
+	mod := make(map[string]string)
+	for fileName, patch := range linkerPatches {
+		oldPath := filepath.Join(srcDirectory, fileName)
+		newPath := filepath.Join(workingDirectory, fileName)
+		mod[oldPath] = newPath
+
+		if err := copyFile(oldPath, newPath); err != nil {
+			return nil, err
+		}
+
+		if err := applyPatch(workingDirectory, patch); err != nil {
+			return nil, fmt.Errorf("apply patch for %s failed: %v", fileName, err)
+		}
+	}
+	return mod, nil
+}
+
+func cachePath(goExe string) (string, error) {
+	var cacheDir string
+	if val, ok := os.LookupEnv(garbleCacheDir); ok {
+		cacheDir = val
+	} else {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			panic(fmt.Errorf("cannot retreive user cache directory: %v", err))
+		}
+		cacheDir = userCacheDir
+	}
+
+	cacheDir = filepath.Join(cacheDir, cacheDirName)
+	if err := os.MkdirAll(cacheDir, os.ModePerm); err != nil {
+		return "", err
+	}
+
+	return filepath.Join(cacheDir, "link"+goExe), nil
```

Comment:
I would add a comment here, like:

Note that we only keep one patched and built linker in the cache.
If the user switches between Go versions or garble versions often,
this may result in rebuilds since we don't keep multiple binaries in the cache.
We can consider keeping multiple versions of the binary in our cache in the future,
similar to how GOCACHE works with multiple built versions of the same package.